overlord/hookstate/ctlcmd/ctlcmd.go

```diff
@@ -91,20 +91,43 @@ func addCommand(name, shortHelp, longHelp string, generator func() command) {
 	}
 }
 
+type ForbiddenCommandError struct {
+	Message string
+}
+
+func (f ForbiddenCommandError) Error() string {
+	return f.Message
+}
+
+type ForbiddenCommand struct {
+	Uid  uint32
+	Name string
+}
+
+func (f *ForbiddenCommand) Execute(args []string) error {
+	return &ForbiddenCommandError{Message: fmt.Sprintf("cannot use %q with uid %d, try with sudo", f.Name, f.Uid)}
+}
+
 // Run runs the requested command.
-func Run(context *hookstate.Context, args []string) (stdout, stderr []byte, err error) {
+func Run(context *hookstate.Context, args []string, uid uint32) (stdout, stderr []byte, err error) {
 	parser := flags.NewParser(nil, flags.PassDoubleDash|flags.HelpFlag)
 
 	// Create stdout/stderr buffers, and make sure commands use them.
 	var stdoutBuffer bytes.Buffer
 	var stderrBuffer bytes.Buffer
 	for name, cmdInfo := range commands {
-		cmd := cmdInfo.generator()
-		cmd.setStdout(&stdoutBuffer)
-		cmd.setStderr(&stderrBuffer)
-		cmd.setContext(context)
-
-		_, err = parser.AddCommand(name, cmdInfo.shortHelp, cmdInfo.longHelp, cmd)
+		var data interface{}
+		// commands listed here will be allowed for regular users
+		if uid == 0 || name == "get" {
```

Comment:
Maybe swap this condition around so that it is easy to see what triggers the forbidden command behavior.

Reply:
Do you mean something like
` if uid != 0 && name == "set" { return &ForbiddenCommand{....} } else { ... }` 

If that's what you mean, I don't think it's a good idea, it becomes easier to forget about updating these restrictions when adding a new command. With the above costruct you need to whitelist commands for regular users, non-listed ones are restricted by default.